### DISCUSSION TITLE
how to handle dangling symlinks while `makepkg`

### DISCUSSION THREAD

**Discussion by goyalyashpal** [msys2/MSYS2-packages#3685] [Q&A] [answered]
- i have uncommented the `MSYS=winsymlinks:nativestrict` line in `msys2.ini` 
(ref: https://github.com/msys2/MSYS2-packages/issues/249#issuecomment-102989479)

- i have enabled "Developer Mode" in Windows settings app (Update & Sec > For devs) 
(ref: https://dev.to/hakonhagland/handling-of-symlinks-on-windows-perl-msys2-cygwin-52h3#symlinks-on-windows-msys2-and-cygwin)

+ i am on `OS Version:                10.0.19042 N/A Build 19042` 20H2
+ i used the default provided `MSYS2 MSYS` shortcut to launch the shell
+ i have successfully installed base-devel via `pacman -Syu base-devel`

On running <tt>[makepkg -CLf]</tt> for the <tt>[PKGBUILD]</tt> file (took from arch) for kakoune, following error is shown:

```console
$ makepkg -CLf
...
==> Extracting sources...
  -> Extracting kakoune-2022.10.31.tar.bz2 with bsdtar
kakoune-2022.10.31/CHANGELOG: Can't create 'kakoune-2022.10.31/CHANGELOG': No such file or directory
kakoune-2022.10.31/libexec/kak/kak: Can't create 'kakoune-2022.10.31/libexec/kak/kak': No such file or directory
bsdtar: Error exit delayed from previous errors.
==> ERROR: Failed to extract kakoune-2022.10.31.tar.bz2
    Aborting...
```

Looking at `CHANGELOG` & `libexec/kak/kak` in the [kakoune repo], both of these have this arrowed-folder icon & the commit message mentions symlink for the latter.

![image](https://user-images.githubusercontent.com/19423063/227537682-e1220f83-398f-4619-a67a-e913352d1ff3.png)
![image](https://user-images.githubusercontent.com/19423063/227537698-8f9d3f39-9cc8-4eff-8d3a-5338383bf1db.png)

[makepkg -CLf]: https://www.msys2.org/wiki/Creating-Packages/#building
[PKGBUILD]: https://github.com/archlinux/svntogit-community/tree/packages/kakoune/trunk
[kakoune repo]: https://github.com/mawww/kakoune/tree/master

**Comment by sskras** [comment]
Does your user have required privileges during the run?

![image](https://user-images.githubusercontent.com/7887758/227548228-e7afe8b5-4fe9-42f0-93f3-3d5431835b89.png)

**Reply by goyalyashpal** [reply]
i checked, and this privilege ain't even shown for the normal user :thinking: - so, i launched the launcher with admin privileges, & while now it showed the privilege, but the state is still showed disabled.

I will look into how to play with those.

```console
$ C:/Windows/System32/whoami.exe //priv

PRIVILEGES INFORMATION
----------------------

Privilege Name                Description                          State
============================= ==================================== ========
SeShutdownPrivilege           Shut down the system                 Disabled
SeChangeNotifyPrivilege       Bypass traverse checking             Enabled
SeUndockPrivilege             Remove computer from docking station Disabled
SeIncreaseWorkingSetPrivilege Increase a process working set       Disabled
SeTimeZonePrivilege           Change the time zone                 Disabled

# C:/Windows/System32/whoami.exe //priv

PRIVILEGES INFORMATION
----------------------

Privilege Name                            Description                                                        State
========================================= ================================================================== ========
SeLockMemoryPrivilege                     Lock pages in memory                                               Disabled
.... 21 MORE HIDDEN ....
SeTimeZonePrivilege                       Change the time zone                                               Disabled
SeCreateSymbolicLinkPrivilege             Create symbolic links                                              Disabled
SeDelegateSessionUserImpersonatePrivilege Obtain an impersonation token for another user in the same session Disabled
```

**Reply by sskras** [reply]
`State` displayed by `whoami.exe` indicates **not** the presence of privilege's grant **but** whether the current process has asked the privilege to be dynamically enabled (in programmatic way) already:
https://social.technet.microsoft.com/Forums/ie/en-US/e24a35b3-fb72-4918-8e51-562e2ad8d8f5/what-is-the-state-column-returned-by-whoami-priv?forum=winserversecurity

... which your `whoami` did not (since it doesn't want to create any symlinks, even if it would be able to do so).

On my machine it's also `Disabled`, but the symlinking works just fine.  And so it should do for you:

![image](https://user-images.githubusercontent.com/7887758/227566170-023fac7b-58f5-4667-b5ac-78d50a297e34.png)

Only that I am unsure if the `bsdtar` extraction failure is related to symlinking.  There might be other reasons.

**Reply by goyalyashpal** [reply]
> Only that I am unsure if the `bsdtar` extraction failure is related to symlinking. There might be other reasons.

i thought same, so, i tried with `tar` with admin privileges, and thankfuly, it explicitly says that it's due to symlinks:

```console
# tar -xf kakoune-2022.10.31.tar.bz2
tar: kakoune-2022.10.31/libexec/kak/kak: Cannot create symlink to ‘../../src/kak’: No such file or directory
tar: Exiting with failure status due to previous errors
```

but the symlink to changelog _is_ created....

wait, lemme test some more.

update: see newest created thread below

**Reply by goyalyashpal** [reply]
just a note, `ll -d` which i guess stands forr `ls -ld` is not a good way to test the symlinks - it shows false positives (where with positive, i mean working/correct/good) _all the time_.

`file` is a _better_ way, but still not reliable. `stat` 's options and aim seem to be not aimed at this

the surefire way to check for a symlink is to check the "location it points to" - either in windows explorer file properties panel, or in some other file explorer (like i did in double commander) or via powershell.

i couldn't find any way to get these info in the direct shell.

**Reply by sskras** [reply]
@goyalyashpal:

> `ll -d` which i guess stands forr `ls -ld` is not a good way to test the symlinks - it shows false positives [...] _all the time_.
> 
> `file` is a _better_ way, but still not reliable. `stat` 's options and aim seem to be not aimed at this

I am surprised.  Do you have some examples of false positives for each way?

**Reply by goyalyashpal** [reply]
> Do you have some examples of false positives for each way?

yeah, i have already documented these at least for the `file` command - the ls and stat were tooo unreliable for me to even consider documenting them. 

refer the table given here: https://github.com/msys2/MSYS2-packages/discussions/3685#discussioncomment-5426436

the 2nd row (red crossed arrow) - for the file of `changelog` under `winsymlinks:native` - the shown output underneath is from `file` - notice that it shows no indication of broken symlink here for changelog.

and shown further down at https://github.com/msys2/MSYS2-packages/discussions/3685#discussioncomment-5429440 is the output to these files from powershell's command - notice that the target column contains `{}` (empty) for this file

**Reply by sskras** [reply]
You essentially flooded this thread, and I am having difficulty on how exactly could one reproduce the results.

Is running `bsdtar -xf ../kakoune-2022.10.31.tar.bz2` and `tar -xf ../kakoune-2022.10.31.tar.bz2` in the empty dirs enough?

With `winsymlinks:nativestrict` neither of both utilities created any of these two symlinks:
```
$ echo $MSYS
winsymlinks:nativestrict

$ mkdir -v kakoune-dir-bsdtar
mkdir: created directory 'kakoune-dir-bsdtar'

$ mkdir -v kakoune-dir-gnutar
mkdir: created directory 'kakoune-dir-gnutar'


$ cd kakoune-dir-bsdtar/

$ bsdtar -xf ../kakoune-2022.10.31.tar.bz2
kakoune-2022.10.31/CHANGELOG: Can't create 'kakoune-2022.10.31/CHANGELOG': No such file or directory
kakoune-2022.10.31/libexec/kak/kak: Can't create 'kakoune-2022.10.31/libexec/kak/kak': No such file or directory
bsdtar: Error exit delayed from previous errors.

$ cd ../kakoune-dir-gnutar/

$ tar -xf ../kakoune-2022.10.31.tar.bz2
tar: kakoune-2022.10.31/CHANGELOG: Cannot create symlink to ‘doc/pages/changelog.asciidoc’: No such file or directory
tar: kakoune-2022.10.31/libexec/kak/kak: Cannot create symlink to ‘../../src/kak’: No such file or directory
tar: Exiting with failure status due to previous errors

$ cd ..

$ ls -Al kakoune-dir-{bsdtar,gnutar}/kakoune-2022.10.31/
kakoune-dir-bsdtar/kakoune-2022.10.31/:
total 114
-rw-r--r-- 1 saukrs None  1181 Oct 31 00:05 .cirrus.yml
drwxr-xr-x 1 saukrs None     0 Oct 31 00:05 .github
-rw-r--r-- 1 saukrs None   145 Oct 31 00:05 .gitignore
drwxr-xr-x 1 saukrs None     0 Oct 31 00:05 colors
drwxr-xr-x 1 saukrs None     0 Oct 31 00:05 contrib
-rw-r--r-- 1 saukrs None   776 Oct 31 00:05 CONTRIBUTING
drwxr-xr-x 1 saukrs None     0 Oct 31 00:05 doc
drwxr-xr-x 1 saukrs None     0 Oct 31 00:05 gdb
drwxr-xr-x 1 saukrs None     0 Oct 31 00:05 libexec
-rw-r--r-- 1 saukrs None    72 Oct 31 00:05 Makefile
drwxr-xr-x 1 saukrs None     0 Oct 31 00:05 rc
-rw-r--r-- 1 saukrs None 28557 Oct 31 00:05 README.asciidoc
drwxr-xr-x 1 saukrs None     0 Oct 31 00:05 share
drwxr-xr-x 1 saukrs None     0 Oct 31 00:05 src
drwxr-xr-x 1 saukrs None     0 Oct 31 00:05 test
-rw-r--r-- 1 saukrs None  1212 Oct 31 00:05 UNLICENSE
-rw-r--r-- 1 saukrs None  2271 Oct 31 00:05 VIMTOKAK

kakoune-dir-gnutar/kakoune-2022.10.31/:
total 114
-rw-r--r-- 1 saukrs None  1181 Oct 31 00:05 .cirrus.yml
drwxr-xr-x 1 saukrs None     0 Oct 31 00:05 .github
-rw-r--r-- 1 saukrs None   145 Oct 31 00:05 .gitignore
drwxr-xr-x 1 saukrs None     0 Oct 31 00:05 colors
drwxr-xr-x 1 saukrs None     0 Oct 31 00:05 contrib
-rw-r--r-- 1 saukrs None   776 Oct 31 00:05 CONTRIBUTING
drwxr-xr-x 1 saukrs None     0 Oct 31 00:05 doc
drwxr-xr-x 1 saukrs None     0 Oct 31 00:05 gdb
drwxr-xr-x 1 saukrs None     0 Oct 31 00:05 libexec
-rw-r--r-- 1 saukrs None    72 Oct 31 00:05 Makefile
drwxr-xr-x 1 saukrs None     0 Oct 31 00:05 rc
-rw-r--r-- 1 saukrs None 28557 Oct 31 00:05 README.asciidoc
drwxr-xr-x 1 saukrs None     0 Oct 31 00:05 share
drwxr-xr-x 1 saukrs None     0 Mar 31 20:31 src
drwxr-xr-x 1 saukrs None     0 Oct 31 00:05 test
-rw-r--r-- 1 saukrs None  1212 Oct 31 00:05 UNLICENSE
-rw-r--r-- 1 saukrs None  2271 Oct 31 00:05 VIMTOKAK

$ ls -Al kakoune-dir-{bsdtar,gnutar}/kakoune-2022.10.31/CHANGELOG
ls: cannot access 'kakoune-dir-bsdtar/kakoune-2022.10.31/CHANGELOG': No such file or directory
ls: cannot access 'kakoune-dir-gnutar/kakoune-2022.10.31/CHANGELOG': No such file or directory

$ ls -Al kakoune-dir-{bsdtar,gnutar}/kakoune-2022.10.31/libexec/kak/kak
ls: cannot access 'kakoune-dir-bsdtar/kakoune-2022.10.31/libexec/kak/kak': No such file or directory
ls: cannot access 'kakoune-dir-gnutar/kakoune-2022.10.31/libexec/kak/kak': No such file or directory
```

File Explorer shows no such item on the FS too.
Maybe that's because I use somewhat older version of MSYS2 ?
```
$ uname -a
MSYS_NT-10.0-19044 DESKTOP-O7JE7JE 3.3.6-341.x86_64 2022-10-24 08:25 UTC x86_64 Msys
```

**Reply by goyalyashpal** [reply]
> ou essentially flooded this thread

sorry for that, yes, i haven't yet found a way to properly record the exploratory data....

i was looking for "hiding" my no longer relevant comments (and also marking the unrelevant threads to be collapsed by default), but that feature is not available in discussions yet on github

anyhow, these are the versions i am having: 

```console
$ uname -srv
MSYS_NT-10.0-19042 3.4.6.x86_64 2023-02-15 18:03 UTC

$ tar --version
bsdtar 3.5.2 - libarchive 3.5.2 zlib/1.2.5.f-ipp

$ bsdtar --version
bsdtar 3.6.2 - libarchive 3.6.2 zlib/1.2.13 liblzma/5.4.0 bz2lib/1.0.8 liblz4/1.9.4 libzstd/1.5.2
```

i will try to think of some way to make it clear for you. but don't hold your breath, it may take long... maybe some weeks at minimum

**Reply by sskras** [reply]
In general I am OK with this because it's your thread.

But I am still interested in why our results differs both:

- in the case of `nativestrict` (both symlinks are missing regardless of the used tool for me)
  and
- in the case of `native` (both symlinks are created by the both tools for me).

OK, my next step will be to update my MSYS2 installation:
```
$ pacman -Q msys2-runtime
msys2-runtime 3.3.6-3
```

**Comment by Biswa96** [comment, marked as answer]
Alternatively, you can extract the tar file twice. For example,

```sh
noextract=("foo.tar.xz")

prepare() {
  tar -xf foo.tar.xz || tar -xf foo.tar.xz
}
```

**Reply by goyalyashpal** [reply]
shall i put <tt>[noextract (array)]</tt> between `source=()` and `sha256sums=()` 
and <tt>[prepare()] Function</tt> just before `build(){}`? 

the resulting will become:

`... optdepends=() source=() sha256sums=() noextract=() prepare(){} build(){} check(){} package(){}`


also, does the order even matter functionally?


[noextract (array)]: https://archlinux.org/pacman/PKGBUILD.5.html#:~:text=contain%20whitespace%20characters.-,noextract%20(array),-An%20array%20of
[prepare()]: https://man.archlinux.org/man/PKGBUILD.5#:~:text=user%20calling%20makepkg.-,prepare(),-Function

**Reply by Biswa96** [reply]
> also, does the order even matter functionally?

No, the order does not matter. Though I generally follow the order. You can find some guidelines

* for mingw specific rules https://github.com/msys2/MINGW-packages/tree/master/mingw-w64-PKGBUILD-templates
* general rules https://wiki.archlinux.org/title/Arch_package_guidelines

**Reply by goyalyashpal** [reply]
> ```shell
> tar -xf foo.tar.xz || tar -xf foo.tar.xz
> ```


thanks, i just tried this now (in the admin/nativestrict environment), 
and while this works for the defined symlinks (here, the CHANGELOG one), 
this still doesn't work for the dangling symlinks; 
here, the libexec/kak/kak one, which will come only after completion of `make` in `build() {}`

**Reply by goyalyashpal** [reply]
i resolved to doing this:

```bash
noextract=("kakoune-2022.10.31.tar.bz2")

prepare() {
    cd $srcdir/
    tar -xf kakoune-2022.10.31.tar.bz2
    touch kakoune-2022.10.31/src/kak
    tar -xf kakoune-2022.10.31.tar.bz2
}
```

**Comment by goyalyashpal** [comment]
ohkay, the `MSYS=winsymlinks:native` as mentioned in that article is correct, not the `MSYS=winsymlinks:nativestrict`

~i asked in kakoune irc, and they mention that that symlink is a dangling one (i.e. destination doesn't exist). do you have any experience with those?~

----
just noting for myself, figure out how to:

- [ ] show recipes (like `make --no-silent`) or the executed commands (like `pacboy debug`). 
Maybe 'would've helped a bit if were shown for 

```console
==> Extracting sources...
  -> Extracting kakoune-2022.10.31.tar.bz2 with bsdtar
```

- [x] inspect and verify the resulting symlink files from CLI, as doublecommander's icon ("pointer" and "pointer to cross") do help but GUI is hard

```console
$ file CHANGELOG libexec/kak/kak
CHANGELOG:       symbolic link to doc/pages/changelog.asciidoc
libexec/kak/kak: broken symbolic link to ../../src/kak

$ stat CHANGELOG libexec/kak/kak
```

----
one surprising  🤔 thing is that, 
- there are 2 cases: with `MSYS=winsymlinks:nativstrict` and `MSYS=winsymlinks:native`
- `file` doesn't say broken for CHANGELOG symlink in either case, but
- doublecmd shows the it to be good (green pointer) in `nativestrict` case
![image](https://user-images.githubusercontent.com/19423063/227716950-296b06eb-ed29-407b-834e-e3e07245b04c.png)
- ... and broken (red pointer to cross) with `native` : 
![image](https://user-images.githubusercontent.com/19423063/227715770-0e18d968-11d2-499a-a0cc-958cc85acc29.png)
- bsdtar can't even create CHANGELOG with `nativestrict`

**Reply by goyalyashpal** [reply]
<table>
<th>
  <td><tt>bsdtar -xf</tt></td><td><tt>tar -xf</tt></td>
</th>
<tr>
  <td><tt>nativestrict</tt><br/>(green ptr in doublecmd)<br/><img src="https://user-images.githubusercontent.com/19423063/227716950-296b06eb-ed29-407b-834e-e3e07245b04c.png"/></td>
  <td>
<tt>CHANGELOG:       
cannot open `CHANGELOG' (No such file or directory)</tt><br/><br/>
<tt>libexec/kak/kak: 
cannot open `libexec/kak/kak' (No such file or directory)</tt>
  </td>
  <td>
<tt>CHANGELOG:       
symbolic link to doc/pages/changelog.asciidoc</tt><br/><br/>
<tt>libexec/kak/kak: 
cannot open `libexec/kak/kak' (No such file or directory)</tt>
</tt>
  </td>
</tr>
<tr>
  <td></td>
  <td colspan=2>In <tt>native</tt>, <tt>bsdtar</tt> and <tt>tar</tt> seems same from behaviour of <tt>file</tt> & doublecmd.</td>
</tr>
<tr>
  <td><tt>native</tt><br/>(red broken ptr in doublecmd)<br/><img src="https://user-images.githubusercontent.com/19423063/227715770-0e18d968-11d2-499a-a0cc-958cc85acc29.png"/></td>
  <td>
<tt>CHANGELOG:       
symbolic link to doc/pages/changelog.asciidoc</tt><br/><br/>
<tt>libexec/kak/kak: 
broken symbolic link to ../../src/kak</tt>
  </td>
  <td>
<tt>CHANGELOG:       
symbolic link to doc/pages/changelog.asciidoc</tt><br/><br/>
<tt>libexec/kak/kak: 
broken symbolic link to ../../src/kak</tt>
  </td>
</tr>
</table>

**Reply by goyalyashpal** [reply]
This difference is apparent on looking at the "Properties" panel of these files

TODO: will add screenshots shortly

----
For viewing the properties in command line, after quite research, i found the following syntax for the `powershell` commandlets:

```powershell
Get-Item -Path <string[]> | Format-List -Property <Object[]>

# aliased, & shortcut form
gi <string[]> | fl <Object[]>
```

where `Get-Item` becomes -> `gi` ; `Format-List` -> `fl` ; `-path` & `-property` gets skipped; 

----

<details><summary>Derivation journey, bit-by-bit</summary>
<br/>

adding the [calculated property] : `{$_.Directory.BaseName}` for showing the parent directory, 
further refining with a label: `@{label="Parent";expression={$_.Directory.BaseName}}`

thus results in:

```powershell
gi changelog,.\libexec\kak\kak | fl Name,Mode,Attributes,LinkType,Target,@{label="Parent";expression={$_.Directory.BaseName}}
```

[calculated property]: https://learn.microsoft.com/en-us/powershell/module/microsoft.powershell.utility/select-object#example-11-create-calculated-properties-for-each-inputobject "select-object via Example-7 in New-Item to create SymbolicLink"


doing some restructuring, required output format for table is: 

```powershell
Attributes, Mode, Name, LinkType, Target, @{label="Parent";expression={$_.Directory.BaseName}
```

... putting some `Format-Table` --> `ft`,  & `foreach` --> `%` loop: 

```powershell
 ".\CHANGELOG",".\libexec\kak\kak" | ForEach {gi $_} | ft Attributes, Mode, Name, LinkType, Target, @{label="Parent";expression={$_.Directory.BaseName}}
```

Add some validity checking, and whitespacing for readability:

```powershell
".\CHANGELOG",".\libexec\kak\kak" | ForEach {if (!(Test-Path $_)) {"`r`n"}; gi $_} | ft Attributes, Mode, Name, LinkType, Target, @{label="Parent";expression={$_.Directory.BaseName}
```

Add somth on the likes of brace expansion:

```powershell
"kakoune-2022.10.31"| % {} {"$_-native"} {"$_-nativestrict"} {} | % {} {"$_-tar"} {"$_-bsdtar"} {} | % {} {"$_\CHANGELOG"} {"$_\libexec\kak\kak"} {} | ForEach {if (!(Test-Path $_)) {"`r`n"}; gi $_} | ft Attributes, Mode, Name, LinkType, Target, @{label="Parent";expression={$_.Directory.BaseName}}
```

</details>

----

Clean-up the directories to show correct relative paths:

```powershell
"kakoune-2022.10.31"| % {} {"$_-native"} {"$_-nativestrict"} {} | % {} {"$_-tar"} {"$_-bsdtar"} {} | % {} {"$_\CHANGELOG"} {"$_\libexec\kak\kak"} {} | ForEach {if (!(Test-Path $_)) {"`r`n"}; gi $_} | ft Attributes, Mode, Name, LinkType, Target, @{label="Parent";expression={Resolve-Path $_.Directory -Relative}}
```

The per-directory usage version:

```powershell
"CHANGELOG", ".\libexec\kak\kak" | ForEach {if (!(Test-Path $_)) {"`r`n"}; gi $_} | ft Attributes, Mode, Name, LinkType, Target
```

**Reply by goyalyashpal** [reply]
Done in the comment below
~TODO: will post the updated logs with the added properties of Name,Mode,Attributes for more clarity~

<details>
<summary>Console logs for the Powershell commandlets</summary>
<p>


```powershell
# Empty - hence the "Red Ptr to Cross" in doublecmd
PS \> cd ..\kakoune-2022.10.31-native-bsdtar\
PS \> gi changelog,.\libexec\kak\kak | fl LinkType,Target


LinkType :
Target   : {}

LinkType :
Target   : {}



PS \> cd ..\kakoune-2022.10.31-native-tar\
PS \> gi changelog,.\libexec\kak\kak | fl LinkType,Target


LinkType :
Target   : {}

LinkType :
Target   : {}



```

```powershell
# Only the changelog is pointing somewhere, hence the green pointer in doublecmd
PS \> cd ..\kakoune-2022.10.31-nativestrict-tar\
PS \> gi changelog,.\libexec\kak\kak | fl LinkType,Target


LinkType : SymbolicLink
Target   : {doc\pages\changelog.asciidoc}

gi : Cannot find path 'D:\15_devt\abs\kakoune\kakoune-2022.10.31-nativestrict-tar\libexec\kak\kak' because it does not exist.
At line:1 char:1
+ gi changelog,.\libexec\kak\kak | fl LinkType,Target
+ ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
    + CategoryInfo          : ObjectNotFound: (D:\15_devt\abs\...libexec\kak\kak:String) [Get-Item], ItemNotFoundException
    + FullyQualifiedErrorId : PathNotFound,Microsoft.PowerShell.Commands.GetItemCommand



```

```powershell
PS \> cd ..\kakoune-2022.10.31-nativestrict-bsdtar\
PS \> gi changelog,.\libexec\kak\kak | fl LinkType,Target
gi : Cannot find path 'D:\15_devt\abs\kakoune\kakoune-2022.10.31-nativestrict-bsdtar\changelog' because it does not exist.
At line:1 char:1
+ gi changelog,.\libexec\kak\kak | fl LinkType,Target
+ ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
    + CategoryInfo          : ObjectNotFound: (D:\15_devt\abs\...sdtar\changelog:String) [Get-Item], ItemNotFoundException
    + FullyQualifiedErrorId : PathNotFound,Microsoft.PowerShell.Commands.GetItemCommand

gi : Cannot find path 'D:\15_devt\abs\kakoune\kakoune-2022.10.31-nativestrict-bsdtar\libexec\kak\kak' because it does not exist.
At line:1 char:1
+ gi changelog,.\libexec\kak\kak | fl LinkType,Target
+ ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
    + CategoryInfo          : ObjectNotFound: (D:\15_devt\abs\...libexec\kak\kak:String) [Get-Item], ItemNotFoundException
    + FullyQualifiedErrorId : PathNotFound,Microsoft.PowerShell.Commands.GetItemCommand
```

</p>
</details>

**Reply by goyalyashpal** [reply]
i just tested with msys launcher with admin privileges, and while the `native` one performed same as above (blank symlinks), but for the `MSYS=winsymlinks:nativestrict` ones this time, even the `tar -xf kakoune-2022.10.31.tar.bz2` failed for both including CHANGLOG

**Reply by goyalyashpal** [reply]
It's kinda impressive that in 8 hours, i went from having never used powershell before (never as in just used in 2-3 distant sets of 5-6 minutes) to writing this whole thing on my own reading documentation. waow.

```powershell
PS kakoune> "kakoune-2022.10.31"| % {} {"$_-native"} {"$_-nativestrict"} {} | % {} {"$_-tar"} {"$_-bsdtar"} {} | % {} {"$_\CHANGELOG"} {"$_\lib
exec\kak\kak"} {} | ForEach {if (!(Test-Path $_)) {"`r`n"}; gi $_} | ft Attributes, Mode, Name, LinkType, Target, @{label="Parent";expression={Resolve-Path $_
.Directory -Relative}}

           Attributes Mode   Name      LinkType     Target                         Parent
           ---------- ----   ----      --------     ------                         ------
Archive, ReparsePoint -a---l CHANGELOG              {}                             .\kakoune-2022.10.31-native-tar
Archive, ReparsePoint -a---l kak                    {}                             .\kakoune-2022.10.31-native-tar\libexec\kak
Archive, ReparsePoint -a---l CHANGELOG              {}                             .\kakoune-2022.10.31-native-bsdtar
Archive, ReparsePoint -a---l kak                    {}                             .\kakoune-2022.10.31-native-bsdtar\libexec\kak
Archive, ReparsePoint -a---l CHANGELOG SymbolicLink {doc\pages\changelog.asciidoc} .\kakoune-2022.10.31-nativestrict-tar


gi : Cannot find path 'kakoune\kakoune-2022.10.31-nativestrict-tar\libexec\kak\kak' because it does not exist.
At line:1 char:192
+ ... \kak\kak"} {} | ForEach {if (!(Test-Path $_)) {"`r`n"}; gi $_} | ft A ...
+                                                             ~~~~~
    + CategoryInfo          : ObjectNotFound: (...libexec\kak\kak:String) [Get-Item], ItemNotFoundException
    + FullyQualifiedErrorId : PathNotFound,Microsoft.PowerShell.Commands.GetItemCommand



gi : Cannot find path 'kakoune\kakoune-2022.10.31-nativestrict-bsdtar\CHANGELOG' because it does not exist.
At line:1 char:192
+ ... \kak\kak"} {} | ForEach {if (!(Test-Path $_)) {"`r`n"}; gi $_} | ft A ...
+                                                             ~~~~~
    + CategoryInfo          : ObjectNotFound: (...sdtar\CHANGELOG:String) [Get-Item], ItemNotFoundException
    + FullyQualifiedErrorId : PathNotFound,Microsoft.PowerShell.Commands.GetItemCommand



gi : Cannot find path 'kakoune\kakoune-2022.10.31-nativestrict-bsdtar\libexec\kak\kak' because it does not exist.
At line:1 char:192
+ ... \kak\kak"} {} | ForEach {if (!(Test-Path $_)) {"`r`n"}; gi $_} | ft A ...
+                                                             ~~~~~
    + CategoryInfo          : ObjectNotFound: (...libexec\kak\kak:String) [Get-Item], ItemNotFoundException
    + FullyQualifiedErrorId : PathNotFound,Microsoft.PowerShell.Commands.GetItemCommand
```

**Reply by goyalyashpal** [reply]
so, i tried a hack: creating a dummy `src/kak` file, then a symlink to that, then deleting the dummy file.

```console
# pwd=kakoune\kakoune-2022.10.31
cd src
touch kak
cd ..
cd libexec/kak
powershell
```

```powershell
cd .\libexec\kak\; New-Item -ItemType SymbolicLink -Path ./kak -Target ../../src/kak; cd ../../;
"CHANGELOG", ".\libexec\kak\kak" | ForEach {if (!(Test-Path $_)) {"`r`n"}; gi $_} | ft Attributes, Mode, Name, LinkType, Target
```

<details><summary>It required admin privileges</summary>
<br/>

```powershell
New-Item : Administrator privilege required for this operation.
At line:1 char:1
+ New-Item -ItemType SymbolicLink -Path ./kak -Target ../../src/kak -Fo ...
+ ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
    + CategoryInfo          : PermissionDenied: (D:\15_devt\abs\...2.10.31\s
   rc\kak:String) [New-Item], UnauthorizedAccessException
    + FullyQualifiedErrorId : NewItemSymbolicLinkElevationRequired,Microsoft
   .PowerShell.Commands.NewItemCommand
```

</details>

so, i opened the msys launcher elevated. but still, the check above showed that the symlink created is absolute.... huh :)

```
           Attributes Mode   Name      LinkType     Target
           ---------- ----   ----      --------     ------
Archive, ReparsePoint -a---l CHANGELOG SymbolicLink {doc\pages\changelog.asciidoc}
Archive, ReparsePoint -a---l kak       SymbolicLink {D:\...kakoune\kakoune-2022...
```
